[build] Consolidation of build system infrastructure

ejgallego opened the pull request:
We replace both the dune-based `(coq.theory ...)` and the "legacy"
make-based build systems for .vo into a single setup, based on our own
Dune rule generation using a coq_dune tool.

This requires a bootstrap process to generate the rule files, until
Dune gets support for `(include stdlib_rules.sexp)` (hopefully soon)

By disabling the `(coq.theory ...)` stanza on `theories/dune` we lose
compositional building of .vo libs (for OCaml we are still fine, which
our main use case), this is not really used as of today so we are
fine.

This should supersede #15545 , avoid many troubles that #15220 had
to endure.

PRs for Docker and OPAM that need to be merged after this:
- https://github.com/coq-community/docker-coq/pull/50 (thanks @erikmd !!)
- https://github.com/coq/opam-coq-archive/pull/1923

List of relevant bugs that should be closed by this PR:
- closes #15635
- closes #15482 
- closes #15730
- closes #8362
- closes #15431
- closes #14342
- closes #14194
- closes #12201
- closes #14512
- closes #10901
- closes #11748
- advances #7149
- closes #8860
- closes #16095
- cc #6137 
- closes #6255
- closes #7208
- cc #8524
- closes #14852

Legacy makefiles will be removed in the next commit, closing a few
more bugs hopefully.

Todos are:

- [x] quick and async CI jobs
- [x] rest of of CI / determine obsolete targets
- [x] install for documentation [check Dune 3.0]
- [x] changelog / documentation update
- [x] should we use coqnative? [not for now maybe?]
- [ ] remove the native hackery for loadpaths now that the build system
      can set that up? Likely for a later PR.

At some point some of this code should be moved to a common loadpath
library to compute native target names, and maybe some other stuff.

Some other changes in `Nativelib`:

- We fix incorrect early initialization of `Boot.Env.t`

- We extend the path passed to the ocamlc compiler as to search for
  files in the dir where the .vo is.

  That can be reverted if we pass all the explicit `-nI` paths.

  We could just get rid of `.coq-native`, additionally.

ejgallego commented:
[Note to self:
  - edge+flambda errors due to native-compiler not finding coqlib (!)
  - it is not possible to backport the `Unix.open_process_args_in` API to 4.05 easily so this runs in < 4.08.
]

SkySkimmer commented:
>it is not possible to backport the Unix.open_process_args_in API to 4.05 easily so this runs in < 4.08.

Just use Unix.pipe and Unix.create_process?

SkySkimmer commented:
added a couple commits

ejgallego commented:
> added a couple commits

need to fix native [which is just copying the rules already in Dune] but that won't happen today :S 

Good news is that once native is fixed, the code is cleaned up, the only real thing to do is to call coqdep only once.

ejgallego commented:
Rebased, and used new coqdep api to speed up rule gen; only todo is native, but that is very easy if someone wants to help.

ejgallego commented:
@ppedrot this is looking good, so hopefully we can make it happen for 8.16, all that we are missing is a few tweaks [modulo stuff review finds out]

cc @SkySkimmer @Alizter

gares commented on tools/dune_rule_gen/util.ml:
```diff
@@ -0,0 +1,42 @@
+(* LICENSE NOTE: This file is dually MIT/LGPL 2.1+ licensed. MIT license:
```
Why?

ejgallego replied:
As explained in the main file header, this implementation is taken from Dune's own coq_rules, in fact they are very close, so once Dune has a plugin API we could unify them. This was the plan we discussed a while back.

ejgallego replied:
Oh, for that particular file, the story is a bit more special, we used to have stuff from OCaml's stdlib, now indeed we still have one function from 4.10 Stdlib, so here we have to use this license.

ejgallego replied:
What I wrote before applies to the other files (args, coq_module, coq_rules, etc...)

Zimmi48 replied:
Dually licensed means that users can take the file and decide what license to apply. If there is code from OCaml inside then the file needs to be licensed under LGPL-2.1-only. It *cannot* be dual licensed with the MIT license and it *cannot* be LGPL-2.1+. Furthermore, the copyright header must be preserved from the OCaml source file. Given that the file seems to only contain one function, just keep the header from the OCaml distribution and do not add anything else.

gares replied:
or rewrite that one function, that is clearly the simplest option

ejgallego replied:
Indeed the license is wrong as this is taken from https://github.com/ocaml/ocaml/blob/trunk/stdlib/list.ml , will amend the license which seems even simpler to me.

ejgallego replied:
The mess of licenses here is due to the old tormented history of this coq_dune development.

silene commented on tools/dune_rule_gen/gen_rules.ml:
```diff
@@ -0,0 +1,29 @@
+  with exn ->
+    let bt = Printexc.get_backtrace () in
+    let exn = Printexc.to_string exn in
+    Format.eprintf "[gen_rules] Fatal error: %s@\n%s@\n%!" exn bt
```
Why does the tool return with exit code 0 in case of a fatal error?

ejgallego replied:
That's a bug, thanks!

SkySkimmer commented on tools/dune_rule_gen/coq_rules.ml:
```diff
@@ -0,0 +1,185 @@
+module FlagUtil = struct
+
+  let plugin_list =
+    [ "ltac"
```
this list should really be dynamically generated

ejgallego replied:
Yes, or even better we can try to figure out how to have coqdep use the meta file and get rid of the list altogheter. Code is there, but I don't understand it yet.

ejgallego commented on tools/dune_rule_gen/coq_rules.ml:
```diff
@@ -0,0 +1,185 @@
+  let dir = prefix_path prefix in
+  if not (dir = "") then
+    Format.fprintf fmt "(subdir %s@\n @[" dir;
+  List.iter (Dune_file.Rule.pp fmt) mods;
+  if not (dir = "") then
+    Format.fprintf fmt "@])@\n";
```
Yes, will fix.

ejgallego commented on tools/dune_rule_gen/dune_file.ml:
```diff
@@ -0,0 +1,44 @@
+  type t =
+    { targets : string list
+    ; deps : string list
+    ; action : string
```
I am not so worried about the formatting. Yes, the idea is to converge with the version in the test suite. I was not planning to go that far, but now that you'd done it it makes sense.

ejgallego commented on tools/dune_rule_gen/dep_info.ml:
```diff
@@ -0,0 +1,49 @@
+
+type t = CD.Dep.t list Dep_map.t
+
+(* What a pita OCaml's stdlib missing basic stuff ... *)
+let from_list l =
+  List.fold_left (fun map { CD.Dep_info.name; deps } ->
+      let name = Path.make name in
+      let path = Path.add_extension ~ext:".v" name in
+      Dep_map.add path deps map) Dep_map.empty l
+
```
I don't want to depend on clib in this development, for portability reasons. It is not enforced yet but it will be soon.

ejgallego commented on tools/dune_rule_gen/coq_rules.mli:
```diff
@@ -0,0 +1,40 @@
+   parameter [boot] controls the kind of the current
+   theory. [root_lvl] tells the rule generator how many levels up the
+   root Coq sources are. *)
+val gen_vo_rules :
+  root_lvl:int
+  -> native:bool
+  -> tname:string list
+  -> boot:Boot_type.t
+  -> Dune_file.Rule.t Dir_info.t
```
Yes, the native flag is here because the coqnative support is not in dune master yet (there's a PR), I agree we should support both methods maybe? This one is faster.

ejgallego replied:
Both methods will be supported in the PR.

ejgallego commented on Makefile.dune:
```diff
@@ -96,15 +97,29 @@ help-install:
-voboot:
-	@echo "This target is empty and not needed anymore"
+# This needs to be regenerated always as to correctly track deps
+_build/default/theories_dune: FORCE
```
Because they are actual files? Is it OK to declare phony for real files?

ejgallego commented on Makefile.dune:
```diff
@@ -51,7 +51,8 @@ help:
 	@echo ""
 	@echo "  - fmt:     run ocamlformat on the codebase"
-	@echo "  - ocheck:  build for all supported OCaml versions [requires OPAM]"
+	@echo "  - ocheck:  check for a selection of supported OCaml versions [requires OPAM]"
+	@echo "  - obuild:  build for a selection of supported OCaml versions [requires OPAM]"
```
The check and build Coq for multiple ocaml versions simultaneously.

ejgallego commented on .gitlab-ci.yml:
```diff
@@ -257,9 +257,6 @@ build:base:
   extends: .build-template
   variables:
     COQ_EXTRA_CONF: "-native-compiler yes -coqide opt"
-    # coqdoc for stdlib, until we know how to build it from installed Coq
-    EXTRA_TARGET: "doc-stdlib"
-    EXTRA_INSTALL: "install-doc-stdlib-html install-doc-printable"
```
Yes, that was a bug of the old build system; the CI setup needs quite a bit of work yet in this PR.

ejgallego replied:
Seems to work fine now, but anyways we need to double check the doc deploy in this PR.

ejgallego commented on tools/coqdep/common.ml:
```diff
@@ -444,6 +444,8 @@ let rec treat_file old_dirname old_name =
   let dirname =
     match (old_dirname,new_dirname) with
       | (d, ".") -> d
+      (* EGJA: We should disable this buggy normalization stuff for
+         "./foo -> foo" but it breaks dune coq.theory! *)
```
No, need to think a bit more about how we do coqdep parsing there indeed.

ejgallego commented on tools/dune_rule_gen/coq_rules.ml:
```diff
@@ -0,0 +1,185 @@
+    ; "rtauto"
+    ] |> List.map (Filename.concat "plugins") |> List.map Path.make
+
+  (* XXX: This should go away once we fully use findlib *)
+  let plugin_flags = Util.list_concat_map (fun p -> [Arg.A "-I"; Arg.Path p]) plugin_list
```
I don't know, I guess when we figure out how it works for coqdep we can just drop this. Doesn't have to happen in this PR tho.

ejgallego commented on tools/coqdep/common.ml:
```diff
@@ -65,7 +65,7 @@ let error_meta_file_lacks_field meta_file package field =
 
 let error_cannot_stat s unix_error =
-  Printf.eprintf "%s\n" (error_message unix_error);
+  Printf.eprintf "%s: %s\n" s (error_message unix_error);
```
Yes, whichever comes first is OK to me.

ejgallego commented on tools/dune_rule_gen/coq_module.ml:
```diff
@@ -0,0 +1,36 @@
+
+let mod_to_obj ~ext { name; _ } = name ^ ext
+
+let obj_files ~tname ~native coq_module  =
+  let native_objs = if native then cmxs_of_mod ~tname coq_module else [] in
+  native_objs @
+  [ mod_to_obj coq_module ~ext:".vo"
+  ; mod_to_obj coq_module ~ext:".glob"
+  ; mod_to_obj coq_module ~ext:".vos"
+  ; "." ^ mod_to_obj coq_module ~ext:".aux"
+  ]
```
Yes, tho I still don't understand how the build rules for .vos work.

ejgallego replied:
vio is solved, vos is not in the current makefile system anyways so not for this PR

ejgallego commented on tools/dune_rule_gen/coq_module.ml:
```diff
@@ -0,0 +1,36 @@
+  ; "." ^ mod_to_obj coq_module ~ext:".aux"
+  ]
+
+let pp _ _ = ()
```
Ups.

ejgallego commented:
Analysis of CI:
- quick: the current rule is incomplete, we need to teach dune how to call schedule2vio ; we incorrectly install the vio files and we overflow the artifact cache
- async: we need to register the dependency to the workers in the vo rules, otherwise the build fails
- doc targets: they are fine, coqbot is confused likely by the renaming
- nix: we need to fix the install of the docs [help needed here]
- osx: no idea
- windows: seems broken on its own, but indeed opam packages will need a tweak
- opam job: succeeds but I'm unsure the resulting coq works, we should improve this test

rest of failures are spurious.

SkySkimmer commented:
osx seems to be a lot of
~~~
Error:
Dynlink error: error loading shared library: Dynlink.Error (Dynlink.Cannot_open_dll "Failure(\"dlopen(/Users/runner/work/coq/coq/_install_ci/lib/coq-core/plugins/ltac/ltac_plugin.cmxs, 10): Symbol not found: _camlCString__2\\n  Referenced from: /Users/runner/work/coq/coq/_install_ci/lib/coq-core/plugins/ltac/ltac_plugin.cmxs\\n  Expected in: flat namespace\\n in /Users/runner/work/coq/coq/_install_ci/lib/coq-core/plugins/ltac/ltac_plugin.cmxs\")")
Findlib paths:
/Users/runner/work/coq/coq/_install_ci/lib
/Users/runner/work/coq/coq/_build/install/default/lib
/Users/runner/work/coq/coq/_install_ci/lib
/Users/runner/work/coq/coq/_build/install/default/lib/coq/../coq-core/..
/Users/runner/work/coq/coq/_build/install/default/lib/coq/user-contrib/Ltac2
src/
/Users/runner/.opam/ocaml-base-compiler.4.12.0/lib
~~~

rgrinberg commented on tools/dune_rule_gen/coq_rules.ml:
```diff
@@ -0,0 +1,185 @@
+  List.iter (Dune_file.Rule.pp fmt) mods;
+  if not (dir = "") then
+    Format.fprintf fmt "@])@\n";
+  (* install stanza doesn't respect subdir , is that a bug? *)
```
I'm not sure I understand the issue from the comment, but `subdir` and `install` should work seamlessly

ejgallego replied:
Seems indeed to work, not sure what the issue was, will report back to dune upstream if I find a problem.

ppedrot commented:
@ejgallego is this still planned to go into 8.16?

silene commented on CONTRIBUTING.md:
```diff
@@ -570,7 +570,7 @@ We have a linter that checks a few different things:
   [Style guide](#style-guide) for additional style recommendations.
 - **Documented syntax is up-to-date**.  If you update the grammar, you
-  should run `make -f Makefile.make doc_gram_rsts` to update the
+  should run `make -f Makefile doc_gram_rsts` to update the
```
```suggestion
  should run `make doc_gram_rsts` to update the
```

SkySkimmer commented:
Commit message says "This should supersede coq#15545" but the commit doesn't touch configure, how does this work?

Alizter commented:
I believe the idea is to remove configure too. I am not exactly sure if dune covers everything that configure was doing with the old makefile however.

ejgallego commented:
> @ejgallego is this still planned to go into 8.16?

Up to you, I think it would be good to ship it in 8.16 indeed. I had to fix a bug in mltop as to make build:vio pass, but things look good in my machine. I don't have time to work on this this week tho :/

gares commented:
All this work is very welcome, but integrating it just before the release seems like an unnecessary risk IMO. Is the fix to mltop you mention something that can be independently applied?

ejgallego commented:
> All this work is very welcome, but integrating it just before the release seems like an unnecessary risk IMO.

Yes, there is a tradeoff here, for me risks are not so big, and the benefits are a lot, but as I mentioned this is up to Pierre-Marie I understand. 

> Is the fix to mltop you mention something that can be independently applied?

Yes, it should go on its own PR.

jfehrle commented:
> All this work is very welcome, but integrating it just before the release seems like an unnecessary risk IMO.

I have to agree.

proux01 commented:
I tried to compile the OPAM packages with coq-native and here is the result:
* `make dunestrap COQ_EXTRA_OPT=-split` generates a `theories/dune` with `-native-compiler off`
* so I added a call to configure in coq-stdlib.opam which makes dunestrap generate a `theories/dune` with `-native-compiler on`
* but then the compilation of coq-stdlib fails with errors like
```
File "././NCoq_Init_Notations.native", line 1, characters 5-17:
1 | open Nativevalues
         ^^^^^^^^^^^^
Error: Unbound module Nativevalues
```
* replacing `dune build -p coq-stdlib` with `dune build -p coq-core,coq-stdlib` in coq-stdlib.opam works but that's not very satisfying (since coq will then be compiled twice by opam)

ejgallego commented:
@proux01 thanks for the report, did you use the `-split` option for coq-stdlib (as done in the local opam file)

I did test split build, but not the native case, and indeed I can see why it is wrong now. I will fix ASAP, as when building in -split mode we need to locate the native headers using findlib, obviously!

Note that the split mode is still considered experimental in this PR, the recommended way for opam packagers is ship a single package for now. I will add a new ci entry `pkg:opam:native` that installs `coq-native` too.

proux01 commented:
> did you use the `-split` option for coq-stdlib (as done in the local opam file) ?

Yes, I copied the commands from the OPAM file:
```
make dunestrap COQ_EXTRA_OPT=-split
```

ppedrot commented:
@ejgallego IIUC this PR going to fix the failure of native compilation of opam packages. Regardless of whether this PR gets in in time, we need to find a way to fix it for the 8.16 release. I guess the choice is between 1. writing a reasonably easy standalone fix for coq-native in 8.16, or 2. merging this PR into 8.16. I think we need input on the risks and feasability of both options before taking a decision.

proux01 commented:
@ppedrot note that the coq-native package is currently (master) not broken with the monolitic OPAM package (coq.opam) as shown by https://github.com/coq/coq/pull/15990 which adds a test for this in the CI
As noted [above](https://github.com/coq/coq/pull/15560#issuecomment-1116025448) by @ejgallego what's broken and could be fixed by this are the split OPAM packages (coq-core and coq-stdlib).

ejgallego commented:
Now depends on https://github.com/coq/coq/pull/15991

I need to push a cleanup to its own PR, but other than that, this should be ready modulo docs and overlay for coq-platform.

Still waiting for review/asignee.

Alizter commented:
One thing I've disliked about `Makefile.dune` is the fact that the default target shows the help screen. Would it not be better to have `make` correspond to `make world`? We can add a `make help` for the help screen too.

Currently the help screen is far too long, especially if somebody tried `make` to intend `make world`. If we are intent on keeping the help screen as the default, then we should at least make it less verbose.

Alizter commented on Makefile.dune:
```diff
@@ -23,15 +23,19 @@ HIDE := $(if $(VERBOSE),,@)
 	@echo "  - watch:  build all public binaries and libraries [continuous build]"
 	@echo "  - check:  build all ML files as fast as possible"
 	@echo "  - test-suite: run Coq's test suite [env NJOBS=N to set job parallelism]"
+	@echo "  - dunestap: Generate dune rules for vo files"
```
Or am I confused with dunestrap and dune-stamp?

ejgallego replied:
No, you are correct, it is a typo.

Alizter commented on coq-stdlib.opam:
```diff
@@ -29,11 +29,22 @@ depends: [
+  #
+  # XXX need to run configure as in coq-core, or else dunestrap will
+  # use the default rule in config
+  [ "./configure"
```
I'm confused. I thought configure was not needed by Makefile.dune?

ejgallego replied:
It is optional. If the developer doesn't run configure, we will use some defaults. That's convenient when developing, but doesn't produce correct packages as the prefix is not properly set.

Alizter commented on INSTALL.md:
```diff
@@ -20,6 +20,9 @@ To compile Coq yourself, you need:
 - [OCaml](https://ocaml.org/) (version >= 4.09.0)
   (This version of Coq has been tested up to OCaml 4.13.0)
 
+  The flambda version of the compiler is known to provide good
```
Perhaps some more details here like the fact that you have to install `ocaml-variants` and `option-flambda` to get this working for more recent versions of OCaml?

ejgallego replied:
Sure, tho we risk that kind of work being out of date. Referering to the official opam documentation seems more robust to me.

Alizter commented on INSTALL.md:
```diff
@@ -68,49 +71,33 @@ Nix users can also get all the required dependencies by running:
-See also [`dev/doc/README.md`](dev/doc/README.md).
+    $ ./configure -prefix <install_prefix> $options
+        $ make dunestrap
+        $ dune build -p coq-core,coq-stdlib,coq,coqide-server,coqide
```
Is this supposed to be indented?

ejgallego replied:
Umm, got messed up yes.

Alizter commented on INSTALL.md:
```diff
@@ -68,49 +71,33 @@ Nix users can also get all the required dependencies by running:
 
-Run-time dependencies of native compilation
--------------------------------------------
+You can drop the coqide packages if not needed.
```
nit: CoqIDE rather than coqide

ejgallego replied:
used a quote instead as the packages are lowercase

Alizter commented on CONTRIBUTING.md:
```diff
@@ -986,7 +986,7 @@ team.
   `.mli` files.  You can browse it on [the Coq website][api-doc], or
-  rebuild it locally (`make -f Makefile.dune apidoc`, requires `odoc`
+  rebuild it locally (`make apidoc`, requires `odoc`
   and `dune`).
```
You can hard wrap this a bit tighter.

Alizter commented on tools/dune_rule_gen/coq_module.ml:
```diff
@@ -0,0 +1,110 @@
+
+let prefix_to_dir = String.concat Filename.dir_sep
+
+(* let target_prefix = if String.equal ext ".cmxs" || String.equal ext ".cmi"
```
Keep or remove?

ejgallego replied:
Yes, remove, thanks!

Alizter commented on tools/dune_rule_gen/coq_module.mli:
```diff
@@ -0,0 +1,51 @@
+
+val pp : Format.formatter -> t -> unit
+
+(* Fix this hack *)
```
Is this a TODO for now or for the future? If the latter, add a TODO.

ejgallego replied:
Maybe should be fixed before merge.

Alizter replied:
What kind of fix do you have in mind?

Alizter commented on tools/dune_rule_gen/coq_module.ml:
```diff
@@ -0,0 +1,110 @@
+    Rule_type.Regular { native }
+  | Rule_type.Regular _ as rule -> rule
+
+(* quick woes... *)
```
?

ejgallego replied:
Improved the comment.

Alizter commented on tools/dune_rule_gen/path.ml:
```diff
@@ -0,0 +1,50 @@
+
+let replace_ext ~ext = map ~f:(fun p -> Filename.remove_extension p ^ ext)
+
+(* Coqdep strips files with dirname "." of the basename, so we need
```
Can you open an issue for coqdep?

Alizter commented on tools/dune_rule_gen/dune:
```diff
@@ -0,0 +1,11 @@
+(library
+ (name coq_dune)
+ (modules :standard \ gen_rules)
+ (flags :standard -w +a-40-42)
```
Can we not ignore these warnings and instead fix the code instead?

ejgallego replied:
There are reasonable warnings to ignore.

ejgallego replied:
Actually mabye @SkySkimmer , the thing is that we don't use the OCaml style I use here a lot in Coq, but for this style is critical. I don't remember the call well, but maybe it makes sense to reopen.

Style is a complex matter tho: I'm very happy with the dune style but of course it is my own pref.

Alizter commented on dev/doc/build-system.dune.md:
```diff
@@ -60,6 +52,15 @@ etc... You should not have to modify `dune` files in regular workflow
 
+## The bootstrap process / rule generation
+
+todo
```
This needs to be finished

Alizter commented on dev/doc/build-system.dune.md:
```diff
@@ -218,9 +219,9 @@ enables stronger compiler optimizations.
+`.opam` files will be automatically generated by Dune from the package
 descriptions in the `dune-project` file; see Dune's manual for more
-details.
+details. For now we have disabled this due to some bugs.
```
What is the exact reason we disabled this?

ejgallego replied:
Dune bugs, in particular `dune subst` was broken for source tree with unicode files. We shall re-enable soon, hopefully.

Alizter replied:
This has been fixed, lets generate opam again.

ejgallego replied:
That requires a dune version bump which may have other implications, I suggest we do that in a different PR, this PR is large enough as it is.

Alizter replied:
I've created an issue for it https://github.com/coq/coq/issues/16071

Alizter commented on doc/plugin_tutorial/dune:
```diff
@@ -0,0 +1,10 @@
+; We get a super bizarre error here...
```
Fix?

ejgallego replied:
I don't know what's going on in the first place.

Alizter commented on library/summary.mli:
```diff
@@ -63,7 +63,7 @@ end
     entries.  Thus is has to be recognizable, and handled properly.
    *)
-val declare_ml_modules_summary : (bool * string) list summary_declaration -> unit
+val declare_ml_modules_summary : (string option * string) list summary_declaration -> unit
```
Can you describe the argments in the doc here?

ejgallego replied:
Done in #15991

Alizter commented on stm/asyncTaskQueue.ml:
```diff
@@ -132,6 +132,7 @@ module Make(T : Task) () = struct
         | ( "-I" | "-nI" | "-include" | "-top" | "-topfile" | "-coqlib" | "-exclude-dir"
           | "-color" | "-init-file"
           | "-profile-ltac-cutoff" | "-main-channel" | "-control-channel"
+          | "-native-output-dir" | "-w"
```
What is native-output-dir?

ejgallego replied:
it control where coqc put the cmxs, the default is `.coq-native`

ejgallego replied:
Done in #15991

Alizter commented on test-suite/dune:
```diff
@@ -18,7 +18,6 @@
    ; File that should be promoted.
    misc/universes/all_stdlib.v
    ; Dependencies of the legacy makefile
```
Don't we need to get rid of all of these?

Alizter commented on dev/shim/dune:
```diff
@@ -2,6 +2,7 @@
  (targets coqtop-prelude)
  (deps
   %{bin:coqtop}
+   ; XXX: bug, we are missing the dep on the _install meta file...
```
Is it possible to directly add the dep on the META?

ejgallego replied:
Not sure, the build vs install layout is biting us here maybe.

ejgallego replied:
I think we will keep using the legacy method for now, and we can address moving to the META in a different PR. It is not very urgent as we need to keep the legacy method until we fix dune.

We can add (if its not there yet) `%{pkg:coq-core:META}` as a dune variable, that is the same tho than expanding it and depending on the install location. Maybe we can do better then and add the rules for a `_build/.lib` layout, expanding OCAMLPATH.

Anyways, both concerns for future PR.

Alizter replied:
So to reply to my less informed past self, `%{lib:coq-core:META}` does not work and atm dune can only rely on `(package coq-core)` to build that META.

Alizter reviewed (commented):
Some initial comments

ejgallego commented:
Sounds good @ppedrot

ejgallego commented:
> One thing I've disliked about `Makefile.dune` is the fact that the default target shows the help screen. Would it not be better to have `make` correspond to `make world`? We can add a `make help` for the help screen too.
> 
> Currently the help screen is far too long, especially if somebody tried `make` to intend `make world`. If we are intent on keeping the help screen as the default, then we should at least make it less verbose, and have an opt-in verbose version. Say `make` listing really common targets and `make help` listing all of them in detail.

Interesting; I find the current setup much better, and it is like 20 lines, so not so long.

For a system as complex as Coq, there is not single good default to have `make` perform a good action. Each user profile will want to do different things.

ejgallego commented on .github/workflows/ci-macos.yml:
```diff
@@ -34,20 +34,25 @@ jobs:
           ./configure -prefix "$(pwd)/_install_ci" -warn-error yes -native-compiler no -coqide opt
-          make -j "$NJOBS"
+          make dunestrap
+          dune build -p coq-core,coq-stdlib,coqide-server,coqide,coq
```
On the :dev job, sorry

ejgallego commented:
This is ready for more review, a few nits to solve, but nothing too serious I can see (failures so far are due to overlays from the depending PR)

- vio job needs to trim the .vio files before the artifact upload.

ejgallego commented on Makefile.dune:
```diff
@@ -93,42 +99,64 @@ help-install:
-	@echo "This target is empty and not needed anymore"
+# We regenerate always as to correctly track deps, can do better
+# We do a single call to dune as to avoid races and locking
+_build/default/theories_dune _build/default/ltac2_dune .dune-stamp: FORCE
```
strap is the phony target, stamp the file.

ejgallego commented:
Rest of CI looking good, modulo nix and Windows build that need to adapt to the new makefile.

Alizter commented on Makefile.build:
```diff
@@ -1,389 +0,0 @@
-TIMING_INCLUDE_MEM?=
-# Option for sorting by the memory column
-TIMING_SORT_BY_MEM?=
-# Output file names for timed builds
-TIME_OF_BUILD_FILE               ?= time-of-build.log
-TIME_OF_BUILD_BEFORE_FILE        ?= time-of-build-before.log
-TIME_OF_BUILD_AFTER_FILE         ?= time-of-build-after.log
-TIME_OF_PRETTY_BUILD_FILE        ?= time-of-build-pretty.log
-TIME_OF_PRETTY_BOTH_BUILD_FILE   ?= time-of-build-both.log
-TIME_OF_PRETTY_BUILD_EXTRA_FILES ?= - # also output to the command line
-BEFORE ?=
```
Do we time builds now?

ejgallego replied:
Actually it would be great to get rid of the regexp based timing system we have and use something better, because the current one is a big hack.

The code that I'm removing here just concerns to the stdlib, and was (poorly duplicated) from coq_makefile, where it is still present. We can extend gen_rules to produce a similar timing log if there is demand. Not sure I'd bother, use case is really small, I'd rather implement #11542

SkySkimmer replied:
This time-of-build stuff isn't used in benching AFAIK

ejgallego replied:
We should def bench this, but I removed this indeed under the assumption Gaëtan points out, we are not really using it, and it is actually not very good maintained because we only test the bits that are in coq_makefile, which of course are legit and not scheduled for removal in any way.

Alizter commented on Makefile.build:
```diff
@@ -1,389 +0,0 @@
-DUNEPLUGININCLUDES=$(addprefix -I _build/default/, $(wildcard plugins/*))
-NATIVEINCLUDES=$(addprefix -nI _build/default/, kernel/.kernel.objs/byte)
-
-ifdef NATIVECOMPUTE
-  COQOPTS += -w -deprecated-native-compiler-option -native-compiler ondemand
-endif
```
What's the correct flag to now pass for native?

Alizter replied:
The reason I am asking is because the native build was one of the sore spots of the dune transition. We need to make sure that absolutely everything done before is supported.

ejgallego replied:
Not sure what you mean. `-native-compiler foo` is deprecated since the introduction of `coqnative`, however that method is much slower.

ejgallego replied:
Yes as of now it is not removed, and given a 33% increase of build time, we may need to tweak coqnative more before we go ahead.

Alizter commented on Makefile.build:
```diff
@@ -1,389 +0,0 @@
-  VO_TOOLS_DEP += $(COQNATIVE)
-endif
-
-ifdef VALIDATE
- VO_TOOLS_DEP += $(CHICKEN)
-endif
```
Validate job?

Alizter replied:
The makefile used to validate by accumulating all the vo files. I have a dune stanza that is able to do that (but for dune 3.0 or before but a bit more painful).

The other solution is to do a new incremental validation (producing a chk log or something). There are flags we can pass to coqchk to allow for this, and worries about the coqchk binary or vo files being swapped out shouldn't hold since the rules will depend on those anyway.

I would suggest the first, and then we can work on the second at a later date.

ejgallego replied:
There are indeed several ways to do validate, but for this PR, is the way we are doing it bad for you?

Alizter replied:
AFAICT we have no validate job anymore.

ejgallego replied:
Yes we do, unless I'm looking to the wrong CI

Alizter replied:
I think there is some confusion. There are two "validate" jobs. One is far calling coqchk on vo files, the other is turning a vio to a vok file. The CI is only doing the latter, the coqchk target has been removed.

Alizter replied:
No you are right, it is there in the CI. The makefile target is still missing however.

SkySkimmer replied:
CI doesn't use the makefiles to call coqchk
see https://gitlab.com/coq/coq/-/jobs/2506489738#L427

Alizter replied:
Yes, I see that now. I guess removing this is OK then. I'll create an issue for adding a build rule for coqchk in the future.

SkySkimmer replied:
IMO we should still provide a validate target, and since it would operate on _build/install we could make CI use it too.

Alizter replied:
@SkySkimmer I've created https://github.com/coq/coq/issues/16072 and I am fairly certain we can replicate both kinds in dune.

ejgallego commented on tools/flock/coq_flock.ml:
```diff
@@ -1,78 +0,0 @@
-(* Simple OCaml version of flock *)
-
-let fail msg =
```
Is not needed anymore.

ejgallego commented on tools/coqdep/lib/common.ml:
```diff
@@ -379,6 +379,8 @@ let rec treat_file old_dirname old_name =
   let dirname =
     match (old_dirname,new_dirname) with
       | (d, ".") -> d
+      (* EGJA: We should disable this buggy normalization stuff for
+         "./foo -> foo" but it breaks dune coq.theory! *)
```
Nope, still I'm not sure what should be done here.

Alizter replied:
If coq.theory gets confused by `./foo` -> `foo` then that is a bug in coq.theory.

ejgallego replied:
Maybe, module vs filenames is actually very subtle (see how `dune coq top` is broken for example)

proux01 commented on .gitlab-ci.yml:
```diff
@@ -322,26 +305,7 @@ lint:
-  script:
-    - opam install coq-native
-    - opam pin add --locked=docker --kind=path coq.dev .
-    # Check that native works
-    - echo "Definition f x := x + x." > test_native.v
-    - coqc -native-compiler yes test_native.v
-    - test -f .coq-native/Ntest_native.cmxs
```
Please keep this test, otherwise the CI may fail to identify failures with coq-native.

ejgallego replied:
Done.

proux01 commented on coq-stdlib.opam:
```diff
@@ -29,11 +29,22 @@ depends: [
+  # use the default rule in config
+  [ "./configure"
+    "-prefix" prefix
+    "-mandir" man
+    "-libdir" "%{lib}%/coq"
+    "-coqide" "no"
+    "-native-compiler" "yes" {coq-native:installed} "no" {!coq-native:installed}
+  ]
+  [ make "dunestrap" "COQ_DUNE_EXTRA_OPT=-split" ]
   [
     "dune"
     "build"
     "-p"
-    name
+    "coq-stdlib"
```
Is there a good reason for that change? Should it also be done in coq-core.opam?

proux01 commented on coq.opam.docker:
```diff
@@ -1,42 +0,0 @@
-synopsis: "The Coq Proof Assistant"
```
Please coordinate with the maintainer of Docker images @erikmd before merging that.

proux01 reviewed (commented):
I retested this with coq-native and it seems to work properly, congrats!

SkySkimmer commented on .gitlab-ci.yml:
```diff
@@ -358,6 +322,16 @@ pkg:opam-split:
+  before_script:
+    - opam install -y coq-native
+  after_script:
+    - echo ""
```
??

proux01 replied:
Seems related to my comment above: https://github.com/coq/coq/pull/15560#discussion_r884241937

ejgallego replied:
Yes, I think I left this as a placeholder for adding that test.

gares commented on tools/dune_rule_gen/LICENSE:
```diff
@@ -0,0 +1,23 @@
+The MIT License
```
Why this exception? All the code is LGPL, no?

ejgallego replied:
This is derived from Dune's code, so IMHO it is better we keep the same license.

Zimmi48 replied:
It should also ease moving code back and forth between Dune's code and the local coq_dune.

ejgallego commented:
> Why the Ltac2/dune file is "disabled"?

Both `theories/dune` and `ltac2/dune` files are not used anymore, as we generate them. I'm not sure if we should just remove the old ones, or keep them around for people that would like to build these libs using the `(coq.theory ...)` support, as coq-universe does.

proux01 commented:
My undestanding was that you intend to "reenable" them at some point in the future, am I wrong?

Alizter commented on CONTRIBUTING.md:
```diff
@@ -1000,15 +999,11 @@ The list of dependencies can be found in the first section of the
 
 Today, the recommended method for building Coq is to use `dune`.  Run
-`make -f Makefile.dune` to get help on the various available targets,
+`make` to get help on the various available targets,
```
```suggestion
`make` to get help on the various available targets.
```

Alizter commented:
Doc artifacts seem to be broken. The rest looks fine to me. Let's get this PR sparkly clean.

ejgallego commented:
Doc artifacts are broken just because we renamed the job, the artifacts themselves are fine. I am not sure how to update coqbot.

Nix needs of course to adapt the install phase, I'll try something but I'd be good if expert Nix users used `build-support/ocaml/dune`.

Windows platform still needs to pin the correct opam files.

Alizter commented:
I can patch coqbot when I'm back at my computer. I've created an issue in the coqbot repo.

ejgallego commented:
Great! Nix seems to work, all we need is someone making the Coq Platform pin the opam packages in this repos.

I won't be able to do this soon unfortunately.

gares commented on .gitlab-ci.yml:
```diff
@@ -76,11 +76,11 @@ before_script:
       - config/coq_config.py
       - config/coq_config.ml
       - config/dune.c_flags
-      - dmesg.txt
```
this is still removed for no reason

SkySkimmer replied:
Has it ever been useful?

gares replied:
nope, at least that I know of. But the other jobs still generate the file... either you don't generate it, or you also keep it

ejgallego replied:
I think it is kept, what job is generating it and not keeping it?

Alizter replied:
You've removed the artefact from Gitlab here, so therefore it is not kept? I think you should revert this removal and leave the cleanup to the future.

Zimmi48 commented on default.nix:
```diff
@@ -86,12 +86,25 @@ stdenv.mkDerivation rec {
   prefixKey = "-prefix ";
 
-  enableParallelBuilding = true;
+  enableParallelBuilding = false;
```
Is that still necessary?

Alizter replied:
The guard against parallel jobs is now in the makefile, so I guess Nix does not need to care anymore.

Zimmi48 replied:
Then, it's better to put this back to `true` just in case Nix puts additional guards that would prevent using multiple cores (when Dune could actually take advantage of them).

Alizter replied:
I think this was intended to stop `make -j4` which ends up calling 4 instances of dune, each with 4 jobs. That also doesn't work since dune will tread on its own feet at the moment. I think Emilio added a guard in the makefile, so even if Nix tries to pass -j4, it will not work.

Zimmi48 replied:
I know what it was intended for since I was the one who recommended it.

Zimmi48 reviewed (commented):
It looks like `make check` changed its meaning (to just running `dune build @check` instead of running the test-suite and then running whatever additional things that `make check` used to run). The value of `installCheckTarget` in `default.nix` should be changed to whatever new target is appropriate (possibly just `test-suite`).

Alizter commented:
So previously the refman artefacts are found in `_build/install` however now they are not present. We probably don't need to change anything for coqbot, just "install" the refman again.

ejgallego commented:
> So previously the refman artefacts are found in `_build/install` however now they are not present. We probably don't need to change anything for coqbot, just "install" the refman again.

Can you share the pointer to where this path is defined in coqbot?

Alizter commented:
@ejgallego  https://github.com/coq/bot/blob/3caa3502a6f397509f1886a177dae155895b18ef/src/actions.ml#L238-L266

ejgallego commented:
Bench is broken with this PR , see https://gitlab.com/coq/coq/-/jobs/2619861726

The error is very strange (and the pin works locally)

Maybe the OPAM version on the bench machine is too old? @SkySkimmer , can you update it?

Alizter commented:
```
[ERROR] The compilation of coq-stdlib failed at "/home/gitlab-runner/builds/y9yFkek2/0/coq/coq/dev/bench/wrapper.sh make dunestrap COQ_DUNE_EXTRA_OPT=-split".

#=== ERROR while compiling coq-stdlib.dev =====================================#
# context              2.0.3 | linux/x86_64 | ocaml-base-compiler.4.09.1 | pinned(file:///home/gitlab-runner/builds/y9yFkek2/0/coq/coq/_bench/coq)
# path                 ~/builds/y9yFkek2/0/coq/coq/_bench/opam.NEW/ocaml-base-compiler.4.09.1/.opam-switch/build/coq-stdlib.dev
# command              ~/builds/y9yFkek2/0/coq/coq/dev/bench/wrapper.sh make dunestrap COQ_DUNE_EXTRA_OPT=-split
# exit-code            2
# env-file             ~/builds/y9yFkek2/0/coq/coq/_bench/opam.NEW/log/coq-stdlib-5012-d2d111.env
# output-file          ~/builds/y9yFkek2/0/coq/coq/_bench/opam.NEW/log/coq-stdlib-5012-d2d111.out
### output ###
# dune build  theories_dune ltac2_dune
# Entering directory '/home/gitlab-runner/builds/y9yFkek2/0/coq/coq'
# Error: Don't know how to build theories_dune
# make: *** [Makefile.dune:114: .dune-stamp] Error 1

[NOTE] Pinning command successful, but your installed packages may be out of sync.
```

Alizter commented:
@ejgallego Try building with opam using a clean git and cache disabled. It looks like the bootstrapping didn't work.

SkySkimmer commented:
Sadly I lost access to the bench machines (some ssh bug maybe), I need to figure out how to get it back.

ejgallego commented:
> > @ejgallego Try building with opam using a clean git and cache disabled. It looks like the bootstrapping didn't work.
> 
> As I just said, it works for me, but I'm using a less broken Opam than the bench. Note how the CI works for OPAM also pass.

But it is a good idea to test with the opam version in the bench machine instead! Let's see!

> Sadly I lost access to the bench machines (some ssh bug maybe), I need to figure out how to get it back.

Oh too bad :/ Ummm, I'll patch bench.sh so it puts a good OPAM, if that's the problem.

ejgallego commented:
> But it is a good idea to test with the opam version in the bench machine instead! Let's see!

In my machine, with opam 2.0.3 as in the bench, things work, damn ! @Alizter can you reproduce locally ?

Alizter commented:
@ejgallego Can you patch `--verbose --debug` into the opam called in the bench script? That way we can get a better description of the failure. You can even do `DUNEOPTS=--verbose` so that we see that dune is doing too.

ejgallego commented:
> @ejgallego Can you patch --verbose --debug into the opam called in the bench script? That way we can get a better description of the failure. You can even do DUNEOPTS=--verbose so that we see that dune is doing too.

Great idea, will do.

Gonna try the bench with the new monkey patch first tho.

I hope this is not a sandbox problem.

ejgallego commented:
I have no idea why coqbot fails to find the refman, it is there.

ejgallego commented:
Any volunteer to fix the platform to actually stop using coq.dev opam files and instead pin to the ones in Coq's repos?

Alizter commented:
> Any volunteer to fix the platform to actually stop using coq.dev opam files and instead pin to the ones in Coq's repos?

I'd suggest we simply disable Windows CI for the time being. We can enable it again when the platform dev scripts are ready. They are unfortunately pretty essential to getting a CI working on windows, so your suggestion would be a lot of work.

Alizter commented:
@ejgallego The refman artefacts are working, I don't see the stdlib doc artefacts however.

ejgallego commented:
> > Any volunteer to fix the platform to actually stop using coq.dev opam files and instead pin to the ones in Coq's repos?
> 
> I'd suggest we simply disable Windows CI for the time being. We can enable it again when the platform dev scripts are ready. They are unfortunately pretty essential to getting a CI working on windows, so your suggestion would be a lot of work. If we do this, let us open an issue so we can remember to fix it.

Umm, that seems too risky, what if the new build system doesn't work on Win?

Maybe today we can have @MSoegtropIMC solve it. It is actually not so much work, just a couple of lines tweak (the hard part is testing it, and the deploy is super horrible, a .zip file]

MSoegtropIMC commented:
@ejgallego : The Coq Platform dev build is there to test in how far a dev build works. This is interesting for me and intentionally as it is and doesn't need any fixing I am aware of.

The point is that Coq CI simply needs something else - likely opam packages patched to ci_basic_overlays.sh. I suggested a while back to create a separate coqci pick which is tailored to the needs of Coq CI. @gares started something like this a long while back (https://github.com/coq/platform/pull/44), but it remained unfinished, after I stated that this is not what I want for Coq Platform CI. We agreed back then to delay any further work on this until the Coq Platform pick system is in place and then do a dedicated coqci pick - which likely would also be quite useful to some users and package maintainers.

So I suggest we discuss what exactly such a coqci pick should do and then we just do it.

ejgallego commented:
> @ejgallego : The Coq Platform dev build is there to test in how far a dev build works. This is interesting for me and intentionally as it is and doesn't need any fixing I am aware of.

The fix needed is that the CI build should pick the opam packages definitions from the canonical source, which is the Coq repos.

This line is not enough, and in fact incorrect:
```
    opam show --raw $pkg.dev | tr -s '\n' ' ' | sed "s@ src: [^ ]*@ src: \"$url\"@" > opam/opam-coq-archive/extra-dev/packages/$pkg/$pkg.dev/opam
```
the opam file should not come from https://coq.inria.fr/opam/core-dev but from the branch the PR is using.

ejgallego commented:
> So I suggest we discuss what exactly such a coqci pick should do and then we just do it.

The problem is that in the first place, the whole coq-core-dev setup is very limited and cannot accommodate any changes we do in Coq upstream without a lot of work.

ejgallego commented:
> @ejgallego The refman artefacts are working, I don't see the stdlib doc artefacts however. They should be here in a html folder:
> 
> https://gitlab.com/coq/coq/-/jobs/2621766528/artifacts/browse/_install_ci/share/doc/coq
> 
> The refman ones are in sphinx.

I see the opposite.

Alizter commented:
> > @ejgallego The refman artefacts are working, I don't see the stdlib doc artefacts however. They should be here in a html folder:
> > [gitlab.com/coq/coq/-/jobs/2621766528/artifacts/browse/_install_ci/share/doc/coq](https://gitlab.com/coq/coq/-/jobs/2621766528/artifacts/browse/_install_ci/share/doc/coq)
> > The refman ones are in sphinx.
> 
> I see the opposite.

In the link I showed you there is one folder called "sphinx". There should be two "sphinx" and "html".

ejgallego commented on tools/dune_rule_gen/path.ml:
```diff
@@ -0,0 +1,50 @@
+
+let relative dir path = map dir ~f:(fun dir -> Filename.concat dir path)
+
+let gen_sub n = String.concat Filename.dir_sep @@ List.init n (fun _ -> "..")
```
> Filename.dir_sep and Filename.concat are just wrong.

Do you mean that a bug should be filled at OCaml upstream?

> On windows `/` is a valid directory separator since forever. Just use that one and not `\` which is well known to cause problems as soon as one interprets a file name as a string with `\` as an escape symbol.

Yes, but we are not interpreting any escaped string here. So that's not a concern. Dune uses the right windows filename separator path and it doesn't have a problem, and that's the recommended setup I am told by Win experts.

It seems to me that it is Coq here who is trying to workaround makefile bugs.

gares replied:
IMO you asked the windows experts the wrong question.

Utilities from `cmd` do use `/` as an option separator, eg `cp -r` is `cp /r`, so passing file names with a `/` to the windows shell is problematic. But passing these names to `coqdep` or any other "unix tool" you call directly (not going trough the cmd shell, that is not via `System.command`) is just fine (including calling bash on cygwin) and the windows syscall to open files is happy with it.

Anyway, the bug here seems that somewhere one uses `/` and somewhere else one uses `\`. I'm suggesting making the normalization *not* platform dependent (or else your dune files generated on windows won't work on linux, which can only serve the purpose of making the testing of this gen_rule harder).

ejgallego replied:
Yes, coqdep uses `/` always, but that's a bug in Coq, see System.// , that was done I guess to avoid forcing proper quoting in the makefiles, but already since a long time we have been using a mix.

That's a bug in coqdep, which has no business overriding OCaml defaults, unless that's a bug in OCaml. If setting the dep to `\` is a bug in OCaml then that's a different issue.

If coqdep users don't have proper quoting that's their problem, not coqdep's.

Also, I don't think we have any business doing path normalization at all, that's slow and serves no purpose, it is again a workaround for lack of a proper spec in terms of what paths are expected along the makefile toolchain as witnessed by #16116

gen_rule works fine on Linux, what it wouldn't work?

Alizter commented:
> > In the link I showed you there is one folder called "sphinx". There should be two "sphinx" and "html".
> 
> How come, that job only builds the refman.

The HTML files genereted by coqdoc that are in the "stdlib refman" are in a different folder.

@Zimmi48 coq bot says:
> doc:refman: stdlib artifact — Link to stdlib build artifact: not found.
[Details](https://gitlab.com/coq/coq/-/jobs/2632689851)
@coqbot-app

When I browse the artefacts:

https://gitlab.com/coq/coq/-/jobs/2632689851/artifacts/browse/_install_ci/share/doc/coq/

I should see a html folder next to sphinx right?

ejgallego commented:
> https://gitlab.com/coq/coq/-/jobs/2632689851/artifacts/browse/_install_ci/share/doc/coq/
> 
> I should see a html folder next to sphinx right?

That job is `doc:refman`, the bot says `   [ ("refman", f "%s/sphinx/html/index.html" url_base) ` and we have that, don't we?

https://coq.gitlab.io/-/coq/-/jobs/2632689851/artifacts/_install_ci/share/doc/coq/sphinx/html/index.html

Alizter commented:
@ejgallego Look at the code for coqbot again. There are two "doc:refman" entries, one for the refman artifact and one for the stdlib artifact: https://github.com/coq/bot/blob/3caa3502a6f397509f1886a177dae155895b18ef/src/actions.ml#L238-L266

ejgallego commented:
Ok, PR submitted https://github.com/coq/bot/pull/232

I think we are good !!

gares commented:
> Yes, rules contain paths with separators, but we don't have a test that compares rules.

don't you think we should have tests for this new gen_rules thing?

ejgallego commented:
> don't you think we should have tests for this new gen_rules thing?

Sure, what kind of tests would you add? It is a simple tool, just generates rules to build the stdlib, it is not meant to be generic.

It is already quite tested tho as basically all CI depends on it, basically the test is to check the files are built properly, and we do that more or less.

Alizter commented:
@ejgallego If we want to add tests, we can have some cram tests, not in test-suite but where gen_rules lives. We can setup something that outputs a few tricky rules for example, that should be easy to setup.

We need to decide what we wish to test however.

ejgallego commented:
Yeah, I'm fine adding tests, but not sure what to test, this program has just a purpose which is to generate the build rules, it either works or not. We didn't test the makefile rules either. If `gen_rules` was a general-purpose build system we could just copy the test suite we have in Dune, but it isn't.

Anyways, I don't think that's a blocker, I'm gonna do a pass on documentation and from my part this is ready to go.

Alizter commented:
OK let's open an issue about tests after this is merged, I am still not 100% sure what we are testing so let's discuss there.

Alizter commented:
## Summary time

1. Since we are now enabling native, some of our CI jobs are taking way longer. Notably fiat-crypto and performance tests.

	@coq/ci-maintainers What should be done about this?

	I suppose we can choose to disable some jobs from using native for the time being, and enable them in the future.

2. The doc:refman: jobs will be fixed by merging coqbot https://github.com/coq/bot/pull/232 after (or before) merging this.

3. Is the bench working @ejgallego?

I would like to merge this today so we can sort out problems before next week starts.

SkySkimmer commented:
>Since we are now enabling native, some of our CI jobs are taking way longer. Notably fiat-crypto and performance tests.

CI already had native, how can that be the change?

Alizter commented:
@SkySkimmer It looks to me like we are calling COQNATIVE on files in this CI but not before.

SkySkimmer commented:
https://gitlab.com/coq/coq/-/jobs/2630919385 (perf tests) has plenty of COQNATIVE

silene commented:
> 2. The doc:refman: jobs will be fixed by merging coqbot [[coq] [doc] Update for new upstream documentation jobs bot#232](https://github.com/coq/bot/pull/232) after (or before) merging this.

Will that break Coqbot for branches other than master, e.g., 8.16?

Alizter commented:
@silene Both the refman artifact and stdlib artifact links won't work like you can see right now. This is because they have been split into separate jobs and had locations changed. The ml api link still works FWIW.

Alizter commented:
The coqbot PR should be backwards compatible and making this CI and other CIs green before we proceed. This is a blocker.

ejgallego commented:
I did a pass on the docs, this is ready to go from my part.

Alizter commented on dev/doc/build-system.dune.md:
```diff
@@ -60,6 +52,52 @@ etc... You should not have to modify `dune` files in regular workflow
+## The bootstrap process / rule generation
+
+Dune is able to build all the OCaml parts of Coq in a pretty standard
+way, using its built-in rule generation for OCaml. Public tTools
```
```suggestion
way, using its built-in rule generation for OCaml. Public tools
```

Alizter commented on dev/doc/build-system.dune.md:
```diff
@@ -60,6 +52,52 @@ etc... You should not have to modify `dune` files in regular workflow
+way, using its built-in rule generation for OCaml. Public tTools
+written in OCaml are distributed in the `coq-core` package.
+
+The set of public `.v` files present in this repos, usually referred
```
```suggestion
The set of public `.v` files present in this repo, usually referred
```

Alizter commented on dev/doc/build-system.dune.md:
```diff
@@ -60,6 +52,52 @@ etc... You should not have to modify `dune` files in regular workflow
+
+The set of public `.v` files present in this repos, usually referred
+as the "Coq Standard Library" are distributed in the `coq-stdlib`
+package. As of June 2022, Dune has to a set of built-in rules for `.v`
```
```suggestion
package. As of June 2022, Dune has a set of built-in rules for `.v`
```

Alizter commented on dev/doc/build-system.dune.md:
```diff
@@ -60,6 +52,52 @@ etc... You should not have to modify `dune` files in regular workflow
+The set of public `.v` files present in this repos, usually referred
+as the "Coq Standard Library" are distributed in the `coq-stdlib`
+package. As of June 2022, Dune has to a set of built-in rules for `.v`
+which is capable of building Coq's standard library fine.
```
```suggestion
files which is capable of building Coq's standard library.
```

Alizter commented on dev/doc/build-system.dune.md:
```diff
@@ -60,6 +52,52 @@ etc... You should not have to modify `dune` files in regular workflow
+
+The rule generation program in `tools/dune_rule_gen` is mostly derived
+from Dune's built-in rules, and it works in an straightforward way: it
+will scan a directory with `.v` files on it, and output the
```
```suggestion
will scan a directory with `.v` files in it, and output the
```

Alizter commented on dev/doc/build-system.dune.md:
```diff
@@ -60,6 +52,52 @@ etc... You should not have to modify `dune` files in regular workflow
+
+In the case of native, the script supports two modes, `coqc
+-native-compiler on` and `coqnative`. The default is the first, as
+currently `coqnative` incurs in a 33% build time overhead in a pretty
```
```suggestion
currently `coqnative` incurs a 33% build time overhead even on a pretty
```

Alizter commented on default.nix:
```diff
@@ -86,12 +86,25 @@ stdenv.mkDerivation rec {
   prefixKey = "-prefix ";
 
-  enableParallelBuilding = true;
+  enableParallelBuilding = false;
```
cc @ejgallego

Alizter replied:
@Zimmi48 asked if this was necessary before.

ejgallego replied:
ok let's revert

Alizter commented on tools/dune_rule_gen/path.ml:
```diff
@@ -0,0 +1,50 @@
+
+let replace_ext ~ext = map ~f:(fun p -> Filename.remove_extension p ^ ext)
+
+(* Coqdep strips files with dirname "." of the basename, so we need
```
@ejgallego could you open an issue about this

ejgallego replied:
This is already discussed in https://github.com/coq/coq/pull/16116 , and that's due to makefile woes. I'm not interesting in fixing it for now.

Alizter replied:
https://github.com/coq/coq/pull/16116

Alizter commented on tools/dune_rule_gen/coq_module.mli:
```diff
@@ -0,0 +1,51 @@
+
+val pp : Format.formatter -> t -> unit
+
+(* Fix this hack *)
```
@ejgallego ping about this too

erikmd commented:
There's an issue with the handling of the `coq-native` package apparently:

Following @ejgallego's suggestion to pre-test the build of `coqorg/coq:dev` in docker-coq before merging this PR, I built the images and inspected one of them (with coq-native):

```bash
$ docker run --rm -it coqorg/coq:newdev-native opam list

# Packages matching: installed
# Name                # Installed # Synopsis
base-bigarray         base
base-threads          base
base-unix             base
conf-gmp              4           Virtual package relying on a GMP lib system installation
coq                   dev         pinned to version dev at git+https://github.com/ejgallego/coq#d6239d7833cdd26a844b02587928a6d3c04abbd7
coq-bignums           dev         Bignums, the Coq library of arbitrary large numbers
coq-core              dev         pinned to version dev at git+https://github.com/ejgallego/coq#d6239d7833cdd26a844b02587928a6d3c04abbd7
coq-native            1           Package flag enabling coq''s native-compiler flag
coq-stdlib            dev         pinned to version dev at git+https://github.com/ejgallego/coq#d6239d7833cdd26a844b02587928a6d3c04abbd7
dune                  3.2.0       pinned to version 3.2.0
num                   1.4         pinned to version 1.4
ocaml                 4.13.1      The OCaml compiler (virtual package)
ocaml-base-compiler   4.13.1      Official release 4.13.1
ocaml-config          2           OCaml Switch Configuration
ocaml-options-vanilla 1           Ensure that OCaml is compiled with no special options enabled
ocamlfind             1.9.1       pinned to version 1.9.1
opam-depext           1.2.1-1     Install OS distribution packages
zarith                1.12        pinned to version 1.12
```

<details>

```bash
# TO BE COMPARED WITH →

$ docker run --rm -it coqorg/coq:dev-native opam list

# Packages matching: installed
# Name                # Installed # Synopsis
base-bigarray         base
base-threads          base
base-unix             base
conf-findutils        1           Virtual package relying on findutils
conf-gmp              4           Virtual package relying on a GMP lib system installation
coq                   dev         pinned to version dev at git+https://github.com/coq/coq#e87d357326ac15980789a80730b6e6105ead9f4e
coq-bignums           dev         Bignums, the Coq library of arbitrary large numbers
coq-native            1           Package flag enabling coq''s native-compiler flag
dune                  3.2.0       pinned to version 3.2.0
num                   1.4         pinned to version 1.4
ocaml                 4.13.1      The OCaml compiler (virtual package)
ocaml-base-compiler   4.13.1      Official release 4.13.1
ocaml-config          2           OCaml Switch Configuration
ocaml-options-vanilla 1           Ensure that OCaml is compiled with no special options enabled
ocamlfind             1.9.1       pinned to version 1.9.1
opam-depext           1.2.1-1     Install OS distribution packages
zarith                1.12        pinned to version 1.12

$ docker run --rm -it coqorg/coq:dev-native find .opam/4.13.1 -name ".coq-native"

.opam/4.13.1/lib/coq/theories/btauto/.coq-native
.opam/4.13.1/lib/coq/theories/NArith/.coq-native
.opam/4.13.1/lib/coq/theories/Classes/.coq-native
.opam/4.13.1/lib/coq/theories/MSets/.coq-native
.opam/4.13.1/lib/coq/theories/ZArith/.coq-native
.opam/4.13.1/lib/coq/theories/ssr/.coq-native
.opam/4.13.1/lib/coq/theories/Array/.coq-native
.opam/4.13.1/lib/coq/theories/Compat/.coq-native
.opam/4.13.1/lib/coq/theories/micromega/.coq-native
.opam/4.13.1/lib/coq/theories/FSets/.coq-native
.opam/4.13.1/lib/coq/theories/Init/.coq-native
.opam/4.13.1/lib/coq/theories/nsatz/.coq-native
.opam/4.13.1/lib/coq/theories/Strings/.coq-native
.opam/4.13.1/lib/coq/theories/ssrmatching/.coq-native
.opam/4.13.1/lib/coq/theories/Bool/.coq-native
.opam/4.13.1/lib/coq/theories/Floats/.coq-native
.opam/4.13.1/lib/coq/theories/funind/.coq-native
.opam/4.13.1/lib/coq/theories/Reals/Abstract/.coq-native
.opam/4.13.1/lib/coq/theories/Reals/.coq-native
.opam/4.13.1/lib/coq/theories/Reals/Cauchy/.coq-native
.opam/4.13.1/lib/coq/theories/Structures/.coq-native
.opam/4.13.1/lib/coq/theories/setoid_ring/.coq-native
.opam/4.13.1/lib/coq/theories/Vectors/.coq-native
.opam/4.13.1/lib/coq/theories/omega/.coq-native
.opam/4.13.1/lib/coq/theories/Sets/.coq-native
.opam/4.13.1/lib/coq/theories/Sorting/.coq-native
.opam/4.13.1/lib/coq/theories/derive/.coq-native
.opam/4.13.1/lib/coq/theories/Logic/.coq-native
.opam/4.13.1/lib/coq/theories/Lists/.coq-native
.opam/4.13.1/lib/coq/theories/rtauto/.coq-native
.opam/4.13.1/lib/coq/theories/QArith/.coq-native
.opam/4.13.1/lib/coq/theories/PArith/.coq-native
.opam/4.13.1/lib/coq/theories/Arith/.coq-native
.opam/4.13.1/lib/coq/theories/Numbers/Cyclic/ZModulo/.coq-native
.opam/4.13.1/lib/coq/theories/Numbers/Cyclic/Abstract/.coq-native
.opam/4.13.1/lib/coq/theories/Numbers/Cyclic/Int63/.coq-native
.opam/4.13.1/lib/coq/theories/Numbers/Cyclic/Int31/.coq-native
.opam/4.13.1/lib/coq/theories/Numbers/Integer/Binary/.coq-native
.opam/4.13.1/lib/coq/theories/Numbers/Integer/Abstract/.coq-native
.opam/4.13.1/lib/coq/theories/Numbers/Integer/NatPairs/.coq-native
.opam/4.13.1/lib/coq/theories/Numbers/NatInt/.coq-native
.opam/4.13.1/lib/coq/theories/Numbers/Natural/Binary/.coq-native
.opam/4.13.1/lib/coq/theories/Numbers/Natural/Peano/.coq-native
.opam/4.13.1/lib/coq/theories/Numbers/Natural/Abstract/.coq-native
.opam/4.13.1/lib/coq/theories/Numbers/.coq-native
.opam/4.13.1/lib/coq/theories/Wellfounded/.coq-native
.opam/4.13.1/lib/coq/theories/extraction/.coq-native
.opam/4.13.1/lib/coq/theories/Setoids/.coq-native
.opam/4.13.1/lib/coq/theories/Program/.coq-native
.opam/4.13.1/lib/coq/theories/Relations/.coq-native
.opam/4.13.1/lib/coq/theories/Unicode/.coq-native
.opam/4.13.1/lib/coq/user-contrib/Ltac2/.coq-native
.opam/4.13.1/lib/coq/user-contrib/Bignums/BigN/.coq-native
.opam/4.13.1/lib/coq/user-contrib/Bignums/CyclicDouble/.coq-native
.opam/4.13.1/lib/coq/user-contrib/Bignums/BigZ/.coq-native
.opam/4.13.1/lib/coq/user-contrib/Bignums/SpecViaZ/.coq-native
.opam/4.13.1/lib/coq/user-contrib/Bignums/BigQ/.coq-native
.opam/4.13.1/lib/coq/user-contrib/Bignums/SpecViaQ/.coq-native
.opam/4.13.1/lib/coq/user-contrib/Bignums/.coq-native

# → OK → e.g. :

$ docker run --rm -it coqorg/coq:dev-native find .opam/4.13.1 -name "NBignums_BigZ_ZMake.cmxs"

.opam/4.13.1/lib/coq/user-contrib/Bignums/BigZ/.coq-native/NBignums_BigZ_ZMake.cmxs
```

</details>

**but**

```bash
$ docker run --rm -it coqorg/coq:newdev-native find .opam/4.13.1 -name ".coq-native"

→ nothing :-/

$ docker run --rm -it coqorg/coq:newdev-native find .opam/4.13.1 -name "NBignums_BigZ_ZMake.cmxs"

→ nothing :-/
```

erikmd commented:
Is it easy to fix this?

would a coq-native compatible dune just require a more recent version of dune?

Alizter commented:
@erikmd The native compatibility is happening in this PR we are generating the rules ourselves, therefore we do not need a newer dune version.

ejgallego commented:
Seems to work on our CI, how can I pull that docker image to see locally?

erikmd commented:
to pull the image, @ejgallego you can do:

`docker pull coqorg/coq:newdev-native`

then run a shell in an ephemeral container with e.g.:

`docker run --rm -it coqorg/coq:newdev-native`

erikmd commented:
> @erikmd The native compatibility is happening in this PR we are generating the rules ourselves, therefore we do not need a newer dune version.

OK maybe… but in this case, the dune version should be bumped in this PR IMHO

ejgallego commented:
> OK maybe… but in this case, the dune version should be bumped in this PR IMHO

Why? We don't need to bump dune for this PR.

erikmd commented:
@ejgallego because before this PR, `coq.dev + coq-native + coq_makfile` works,
and after this PR, a constraint is missing to tell that with the seemingly compatible dune, it does not work

erikmd commented:
but maybe I'm mistaken and this is not the correct explanation/fix, I dunno.

Anyway, the two images I mentioned in my previous comment were the very same docker build of this PR vs. last coq.master

ejgallego commented:
I don't see how that is related. The problem is that in the docker image, we have this
```
$ cat /home/coq/.opam/4.13.1/lib/coq-core/config/coq_config.ml | grep -i native
let native_compiler = NativeOff
```
so the configure step failed.

Alizter commented:
In any case, we need to work out what is happening on docker before merging this.

ejgallego commented:
After you install `coq-native`, is coq-core being rebuilt? That is maybe the problem.

erikmd commented:
> After you install `coq-native`, is coq-core being rebuilt? That is maybe the problem.

Good point, maybe the `coq-core.opam` file is wrong.

erikmd commented:
and yet I see this:

https://github.com/ejgallego/coq/blob/5b491c60dd2048999e45b8a3b3a367a57170d680/coq-core.opam#L44

ejgallego commented:
I dunno what happened in the Docker image, but if I do `opam reinstall coq-core`, native works.

erikmd commented:
And if we start from a non-native image, what you said [here](https://github.com/coq/coq/pull/15560#issuecomment-1165785880) can be reproduced:

```bash

$ docker run --rm -it coqorg/coq:newdev
Unable to find image 'coqorg/coq:newdev' locally
newdev: Pulling from coqorg/coq
c1ad9731b2c7: Already exists 
ce388ca54220: Pull complete 
994ee1df670c: Pull complete 
5da19fd5708e: Pull complete 
fa9a8337e221: Pull complete 
2483dd5d816e: Pull complete 
Digest: sha256:d217b3d19c7d8249d06062ef64dfc203e98682ce556372e174333ae945cf3e4d
Status: Downloaded newer image for coqorg/coq:newdev
coq@c910ea6b3392:~$ opam list
# Packages matching: installed
# Name               # Installed    # Synopsis
base-bigarray        base
base-threads         base
base-unix            base
conf-gmp             4              Virtual package relying on a GMP lib system installation
coq                  dev            pinned to version dev at git+https://github.com/ejgallego/coq#d6239d7833cdd26a844b02587928a6d3c04abbd7
coq-bignums          dev            Bignums, the Coq library of arbitrary large numbers
coq-core             dev            pinned to version dev at git+https://github.com/ejgallego/coq#d6239d7833cdd26a844b02587928a6d3c04abbd7
coq-stdlib           dev            pinned to version dev at git+https://github.com/ejgallego/coq#d6239d7833cdd26a844b02587928a6d3c04abbd7
dune                 3.2.0          pinned to version 3.2.0
num                  1.4            pinned to version 1.4
ocaml                4.13.1         The OCaml compiler (virtual package)
ocaml-config         2              OCaml Switch Configuration
ocaml-option-flambda 1              Set OCaml to be compiled with flambda activated
ocaml-variants       4.13.1+options Official release of OCaml 4.13.1
ocamlfind            1.9.1          pinned to version 1.9.1
opam-depext          1.2.1-1        Install OS distribution packages
zarith               1.12           pinned to version 1.12
coq@c910ea6b3392:~$ opam install coq-native
The following actions will be performed:
  - install coq-native 1

<><> Processing actions <><><><><><><><><><><><><><><><><><><><><><><><><><><><>
-> installed coq-native.1
Done.
```

then coq-core is not rebuilt

Alizter commented on .gitlab-ci.yml:
```diff
@@ -76,11 +76,11 @@ before_script:
       - config/coq_config.py
       - config/coq_config.ml
       - config/dune.c_flags
-      - dmesg.txt
```
@ejgallego  ping. Remove this change and leave for future cleanup. We should remove it when we stop generating it as Enrico suggests.

ejgallego replied:
The diff is misleading, the behavior should be preserved

Alizter commented on default.nix:
```diff
@@ -86,12 +86,28 @@ stdenv.mkDerivation rec {
 
-  installTargets =
-    [ "install" ] ++ optional buildDoc "install-doc-html";
+  # TODO, building of documentation package when not in dev mode
+  # https://github.com/coq/coq/issues/16198
+  # buildFlags = [ "world" ] ++ optional buildDoc "refman-html";
```
Open an issue

ejgallego replied:
It is already open, as noted in the comment

erikmd commented:
@ejgallego actually I think I found the culprit thanks to `opam show --raw …`:

could you please add in the .opam files (maybe all in your PR branch) this line?

```
depopts: ["coq-native"]
```

erikmd commented:
like this:

https://github.com/ocaml/opam-repository/blob/7a3f9165f2941b8d6520c1473d29a7395bc2792e/packages/coq/coq.8.15.2/opam#L20-L22

ejgallego commented:
@erikmd yes, I forgot depopts obviously!!! I always get confused.

Alizter commented:
Good catch @erikmd!

erikmd commented:
> @erikmd yes, I forgot depopts obviously

@ejgallego could you please push another commit that adds

```
depopts: [
  "coq-native"
]
```

in `coq.opam`, `coq-core.opam` and `coq-stdlib.opam` ?

Then, I'll re-trigger the build of `coqorg/coq:newdev-native`, to test everything is OK
before merging [PR coq-community/docker-coq#50](https://github.com/coq-community/docker-coq/pull/50) **then** #155560

ejgallego commented:
Done thanks @erikmd , I added it to `coq-core.opam` only, as I understand that's the only package that checks that value.

`coq-stdlib` will be rebuilt when `coq-core` is. Does my reasoning sound correct?

I've also updated the `dune-project` file, we should start thinking of generating our opam files automatically now that we depend on dune 2.9

ejgallego commented:
> So even if `coq-stdlib` should normally be automatically/indirectly rebuilt (because `coq-stdlib` depends on `coq-core`),
> I'd say it would be safer (and maybe more legible?) to also add the `depopts` in `coq-stdlib.opam`… WDYT?

Yes I think you are right, in the sense that it would be possible for coq-core to actually not depend on coq-native (for example if we would use `coqnative` binary, and then the rules for coq-stdlib would go that way.

coqnative is too slow tho so I didn't enable it by default.

erikmd commented:
@Alizter you removed the label `needs: fixing` *but* IINM the following is still missing:

`"… it would be safer (and maybe more legible?) to also add the depopts in coq-stdlib.opam"`

Alizter commented:
@erikmd I removed the "needs: fixing" label since I added it about the native compilation. I did say I would possibly merge today, however I have decided not to just yet until all these details are fully resolved.

proux01 commented:
> do you still confirm that when `-native-compiler yes` is passed to `./configure`, the generation of `.coq-native/*.cmxs` files behaves the same with this PR w.r.t. before this PR (regarding their availability and their paths) ?

@erikmd : yes, c.f., the `opam:native` job in `.gitlab-ci.yml`

Alizter commented:
Merging after https://github.com/coq-community/docker-coq/pull/50

Alizter commented:
Merging when green

Alizter commented:
@coqbot merge now
